src/httpserver.cpp

```diff
@@ -232,7 +233,14 @@ static void http_request_cb(struct evhttp_request* req, void* arg)
             }
         }
     }
-    std::unique_ptr<HTTPRequest> hreq(new HTTPRequest(req));
+    std::unique_ptr<HTTPRequest> hreq;
+
+    try {
+        hreq = std::make_unique<HTTPRequest>(req);
+    } catch (const std::runtime_error& e) {
+        LogPrint(BCLog::HTTP, "%s\n", e.what());
+        return;
+    }
```

Comment:
I don't think it's great that we're not doing the address-based checks before anything else. Since we're now adding validation into the constructor, how about we do it consistently and do it for all validation, so we can also control what we validate first? I'm still feeling a bit uneasy about coupling the construction of `HTTPRequest` with validation, but if we're doing it we might as well do it more consistently? Curious to hear @vasild's thoughts on this too.

<details>
<summary>git diff</summary>

```diff
diff --git a/src/httpserver.cpp b/src/httpserver.cpp
index f7dc344f4..10f8ec24d 100644
--- a/src/httpserver.cpp
+++ b/src/httpserver.cpp
@@ -235,6 +235,7 @@ static void http_request_cb(struct evhttp_request* req, void* arg)
     }
     std::unique_ptr<HTTPRequest> hreq;
 
+    // Initializing the HTTPRequest also performs basic validation such as address, method and URI checks
     try {
         hreq = std::make_unique<HTTPRequest>(req);
     } catch (const std::runtime_error& e) {
@@ -242,22 +243,6 @@ static void http_request_cb(struct evhttp_request* req, void* arg)
         return;
     }
 
-    // Early address-based allow check
-    if (!ClientAllowed(hreq->GetPeer())) {
-        LogPrint(BCLog::HTTP, "HTTP request from %s rejected: Client network is not allowed RPC access\n",
-                 hreq->GetPeer().ToStringAddrPort());
-        hreq->WriteReply(HTTP_FORBIDDEN);
-        return;
-    }
-
-    // Early reject unknown HTTP methods
-    if (hreq->GetRequestMethod() == HTTPRequest::UNKNOWN) {
-        LogPrint(BCLog::HTTP, "HTTP request from %s rejected: Unknown HTTP request method\n",
-                 hreq->GetPeer().ToStringAddrPort());
-        hreq->WriteReply(HTTP_BAD_METHOD);
-        return;
-    }
-
     const std::string strURI{hreq->GetURI()};
     LogPrint(BCLog::HTTP, "Received a %s request for %s from %s\n",
              RequestMethodString(hreq->GetRequestMethod()), SanitizeString(strURI, SAFE_CHARS_URI).substr(0, 100), hreq->GetPeer().ToStringAddrPort());
@@ -545,13 +530,7 @@ HTTPRequest::HTTPRequest(struct evhttp_request* _req) :
         m_uri{req != nullptr ? evhttp_request_get_uri(req) : nullptr},
         m_uri_parsed{m_uri != nullptr ? evhttp_uri_parse(m_uri) : nullptr}
 {
-    if (m_uri_parsed == nullptr) {
-        const char* err{"URI parsing failed, it likely contained RFC 3986 invalid characters"};
-        const std::string peerAddress{GetPeer().ToStringAddr()};
-        WriteReply(HTTP_BAD_REQUEST, err);
-        throw std::runtime_error(
-            strprintf("HTTP request from %s rejected: %s", peerAddress, err));
-    }
+    Validate();
 }
 
 HTTPRequest::~HTTPRequest()
@@ -565,6 +544,29 @@ HTTPRequest::~HTTPRequest()
     evhttp_uri_free(m_uri_parsed);
 }
 
+void HTTPRequest::Validate()
+{
+    const std::string rejected{strprintf("HTTP request from %s rejected: ", GetPeer().ToStringAddrPort())};
+    // Early address-based allow check
+    if (!ClientAllowed(GetPeer())) {
+        WriteReply(HTTP_FORBIDDEN);
+        throw std::runtime_error(rejected + "Client network is not allowed RPC access");
+    }
+
+    // Early reject unknown HTTP methods
+    if (GetRequestMethod() == HTTPRequest::UNKNOWN) {
+        WriteReply(HTTP_BAD_METHOD);
+        throw std::runtime_error(rejected + "Unknown HTTP request method");
+    }
+
+    // Ensure valid URI
+    if (m_uri_parsed == nullptr) {
+        const std::string err{"URI parsing failed, it likely contained RFC 3986 invalid characters"};
+        WriteReply(HTTP_BAD_REQUEST, err);
+        throw std::runtime_error(rejected + err);
+    }
+}
+
 std::pair<bool, std::string> HTTPRequest::GetHeader(const std::string& hdr) const
 {
     const struct evkeyvalq* headers = evhttp_request_get_input_headers(req);
diff --git a/src/httpserver.h b/src/httpserver.h
index 77a249a6a..3c5af5b0b 100644
--- a/src/httpserver.h
+++ b/src/httpserver.h
@@ -64,6 +64,9 @@ private:
     /// The parsed URI, as returned by `evhttp_uri_parse()`. Will never be `nullptr`.
     evhttp_uri* const m_uri_parsed;
 
+    // Perform early validity checks. Send a response and throw std::runtime_error if invalid.
+    void Validate();
+
 public:
     explicit HTTPRequest(struct evhttp_request* req);
     ~HTTPRequest();

```
</details>

Reply:
I understand your worries on coupling the validation within the constructor (separation of/ different concerns, error handling, flexibility, etc) but since those validations/ "early checks" also would impact into the state of the object itself it would make sense to me to put them together, consistently and cleaner. Happy to make the change if @vasild also agree with it.